### PR TITLE
Execute CI tests in parallel

### LINKED ISSUE #189: test: split up integration tests
When creating a pr it can take quite some time until we know if an... tbd

### PR DESCRIPTION
**Pull request by joschkabraun**
Closes #189.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -167,33 +167,21 @@ jobs:
...
   core-test:
-    needs: prep-testbed
     runs-on: ubuntu-latest
+    needs: update-bff-playground-if-required
     strategy:
       fail-fast: false
       matrix:
         python-version: [3.7]
-        test-path: ${{fromJson(needs.prep-testbed.outputs.matrix)}}
+        group: [1, 2, 3, 4]     # set to the number of end-to-end tests
```
Comment (makram93) [inline .github/workflows/ci.yml:177]: What does this group do? And how are the tests (unit + integration) divided?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -210,38 +198,48 @@ jobs:
...
+      - name: Download all coverage artifacts
+        uses: actions/download-artifact@v2
+      - name: Run coverage
+        run: |
+          coverage combine coverage*/.coverage*
+          coverage xml
+      - name: Upload coverage from test to Codecov
+        uses: codecov/codecov-action@v2
+        with:
+          fail_ci_if_error: false
```
Comment (makram93) [inline .github/workflows/ci.yml:242]:
```suggestion
```
It already defaults to `false` as mentioned [here](https://github.com/codecov/codecov-action#:~:text=Specify%20if%20CI%20pipeline%20should%20fail%20when%20Codecov%20runs%20into%20errors%20during%20upload.%20Defaults%20to%20false) so we may not need to explicitly specify here:

Comment (makram93) [inline .github/workflows/ci.yml:242]:
```suggestion
```
It already defaults to `false` as mentioned [here](https://github.com/codecov/codecov-action#:~:text=Specify%20if%20CI%20pipeline%20should%20fail%20when%20Codecov%20runs%20into%20errors%20during%20upload.%20Defaults%20to%20false) so we may not need to explicitly specify here:

File: requirements-dev.txt [hunk 3]
```diff
@@ -1,2 +1,4 @@
 pytest
 pytest-mock==3.7.0
+pytest-split==0.8.0
+pytest-cov
```
Comment (makram93) [inline requirements-dev.txt:4]: I think we can omit this as this file is nowhere used. In my next PR I will delete this file as we only make use of
`pip install --no-cache-dir ".[full,test]"` in the CI/CD

Comment (makram93) [inline requirements-dev.txt:4]: I think we can omit this as this file is nowhere used. In my next PR I will delete this file as we only make use of
`pip install --no-cache-dir ".[full,test]"` in the CI/CD

File: .test_durations [hunk 4]
```diff
@@ -0,0 +1,94 @@
+{
```
Comment (makram93) [inline .test_durations:1]: Can you also add a note here - 
`When there are major changes in the test cases (execution time) compared to what's stored in this file, it's recommended to update the duration information with --store-durations to ensure that the splitting is in balance.`

`pytest --store-durations`

Comment (makram93) [inline .test_durations:1]: Can you also add a note here - 
`When there are major changes in the test cases (execution time) compared to what's stored in this file, it's recommended to update the duration information with --store-durations to ensure that the splitting is in balance.`

`pytest --store-durations`

**Review by makram93 — COMMENTED** [review]
A very nice PR. I learned about parallel execution today :)
I have left some minor comments.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (joschkabraun) [reply .github/workflows/ci.yml:177]: They are the indexes of groups to run the tests in parallel. The tests are split into the groups such that the time needed to complete them is roughly the same. For that `.test_durations` is used.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (joschkabraun) [reply .github/workflows/ci.yml:242]: Good point. Thank you!

File: .test_durations (hunk 4 above)
Comment (joschkabraun) [reply .test_durations:1]: Good idea! Added it to `tests/README.md`

**Comment by joschkabraun** [comment]
Thank you Kalim for the feedback!

**Review by makram93 — APPROVED** [review]
Once the CI passes, it is good to go!

**Pull request merged**